File: data_structures_java/MyCircularQueue.java
public class MyCircularQueue<T> {
	
	private int front;
	private int rear;
	private int max;
	private T[] arr;

	MyCircularQueue() {
		this(5);
	}
	MyCircularQueue(int max) {
		this.max = max;
		front = -1;
		rear = -1;
		arr = (T[]) new Object[max];
	}
	public boolean isEmpty() {
		return (front == -1 && rear == -1);
	}
	public boolean isFull() {
		return ((rear - front + max)%max == max - 1);
	}
	void enQueue(T item) {
		if (isEmpty()) {
			front = 0;
			rear = 0;
			arr[front] = item;
		} else if (!isFull()){
			rear = ( rear + 1 ) % max;
			arr[rear] = item;
		}
	}
	void deQueue() {
		if (front == rear) {
			front = -1;
			rear = -1;
		} else {
			front = ( front + 1 ) % max;
		}
	}
	void printQueue() {
		if (!isEmpty()){
			int iters = ( rear-front+max ) % max;
			for (int i=0; i<=iters; i++) {
				System.out.println( arr[ (front+i)%max ] );
			}
		}
	}

	public static void main(String[] args) {
		MyCircularQueue<String> mq = new MyCircularQueue<String>();

		mq.enQueue("34");
		mq.enQueue("14");
		mq.enQueue("394");
		mq.enQueue("4");
		mq.enQueue("77");
		mq.enQueue("12");   // won't be added since out of space
	
		mq.deQueue();
		mq.deQueue();
		mq.deQueue();
	
		mq.printQueue();

	}
}

File: data_structures_js/mycircularqueue.js
class MyCircularQueue {
	
	constructor(max = 5) {
		this.max = max;
		this.front = -1;
		this.rear = -1;
		this.arr = [];
	}
	isEmpty() {
		return (this.front == -1 && this.rear == -1);
	}
	isFull() {
		return ((this.rear - this.front + this.max)%this.max == this.max - 1);
	}
	enQueue(item) {
		if (this.isEmpty()) {
			this.front = 0;
			this.rear = 0;
			this.arr[this.front] = item;
		} else if (!this.isFull()){
			this.rear = (this.rear + 1) % this.max
			this.arr[this.rear] = item;
		}
	}
	deQueue() {
		if (this.front == this.rear) {
			this.front = -1;
			this.rear = -1;
		} else {
			this.front = (this.front + 1) % this.max
		}
	}
	printQueue() {
		if (!this.isEmpty()){
			let iters = ( this.rear-this.front+this.max ) % this.max;
			for (let i=0; i<=iters; i++) {
				console.log(
					this.arr[ (this.front+i)%this.max ]
				)
			}
		}
	}
}

if (require.main == module) {

	mq = new MyCircularQueue();

	mq.enQueue("34");
	mq.enQueue("14");
	mq.enQueue("394");
	mq.enQueue("4");
	mq.enQueue("77");
	mq.enQueue("12");   // won't be added since out of space

	mq.deQueue();
	mq.deQueue();
	mq.deQueue();

	mq.printQueue();
}